class.jetpack.php

```diff
@@ -436,56 +436,63 @@ class_exists( 'Jetpack_Sitemap_Manager' )
 		delete_transient( self::$plugin_upgrade_lock_key );
 	}
 
+	/**
+	 * Saves all the currently active modules to options.
+	 * Also fires Action hooks for each newly activated and deactived module.
+	 * 
+	 * @param $modules Array Array of active modules to be saved in options.
+	 * 
+	 * @return $success bool true for success, false for failure.
+	 */
 	static function update_active_modules( $modules ) {
-		$current_modules = Jetpack_Options::get_option( 'active_modules', array() );
-
-		$success = Jetpack_Options::update_option( 'active_modules', array_unique( $modules ) );
-
-		if ( is_array( $modules ) && is_array( $current_modules ) ) {
-			$new_active_modules = array_diff( $modules, $current_modules );
-			foreach( $new_active_modules as $module ) {
-				/**
-				 * Fires when a specific module is activated.
-				 *
-				 * @since 1.9.0
-				 *
-				 * @param string $module Module slug.
-				 * @param boolean $success whether the module was activated. @since 4.2
-				 */
-				do_action( 'jetpack_activate_module', $module, $success );
-
-				/**
-				 * Fires when a module is activated.
-				 * The dynamic part of the filter, $module, is the module slug.
-				 *
-				 * @since 1.9.0
-				 *
-				 * @param string $module Module slug.
-				 */
-				do_action( "jetpack_activate_module_$module", $module );
-			}
-
-			$new_deactive_modules = array_diff( $current_modules, $modules );
-			foreach( $new_deactive_modules as $module ) {
-				/**
-				 * Fired after a module has been deactivated.
-				 *
-				 * @since 4.2.0
-				 *
-				 * @param string $module Module slug.
-				 * @param boolean $success whether the module was deactivated.
-				 */
-				do_action( 'jetpack_deactivate_module', $module, $success );
-				/**
-				 * Fires when a module is deactivated.
-				 * The dynamic part of the filter, $module, is the module slug.
-				 *
-				 * @since 1.9.0
-				 *
-				 * @param string $module Module slug.
-				 */
-				do_action( "jetpack_deactivate_module_$module", $module );
-			}
+		$current_modules      = Jetpack_Options::get_option( 'active_modules', array() );
```

Comment:
I'm not sure I understand why we need 2 different ways of obtaining the currently active modules:

* `Jetpack_Options::get_option( 'active_modules', array() );`
* `Jetpack::get_active_modules()`

I'd expect we only have a single way to do it, ideally with `Jetpack::get_active_modules()`, as it adds some sugar on top of `Jetpack_Options::get_option( 'active_modules', array() );`.

I might be missing something though, so I could use some elaboration on this one.

Reply:
Checking every call of `update_active_modules`‚ the function param `$modules` is always a filtered array value (filtered by `jetpack_active_modules` filter), so it cannot be used directly to persist in the options table. 

What happens when a module is deactivated?
`$modules` will be missing both the deactivated module and the filtered out modules.  We need to [compare it against](https://github.com/Automattic/jetpack/pull/12270/files/23a9976e3220ec4ebae74eebe9881e4016aaf5d1#diff-9ba9fa5bb52bd6cf36c6c0bdd5ed4830R452) `Jetpack::get_active_modules()` to get the exact module(s) that were deactivated.

To update the options table, we need to fetch the `active_modules` option, append or remove modules, then update the options again. It doesn't look like this can be done without a call to `Jetpack_Options::get_option( 'active_modules', array() )`.

Reply:
Brilliant, that makes sense, thanks a lot for explaining 👍